cadquery/occ_impl/solver.py

```diff
@@ -0,0 +1,225 @@
+                    if isinstance(m1, gp_Pnt):
+                        rv += pt_cost(m1, m2, t1, t2, d)
+                    elif isinstance(m1, gp_Dir):
+                        rv += dir_cost(m1, m2, t1, t2, d)
+                    else:
+                        raise NotImplementedError(f"{m1,m2}")
+
+            return rv
+
+        def jac(x):
+
+            constraints = self.constraints
+            ne = self.ne
+
+            delta = DIFF_EPS * eye(NDOF)
+
+            rv = zeros(NDOF * ne)
+
+            transforms = [
+                self._build_transform(*x[NDOF * i : NDOF * (i + 1)]) for i in range(ne)
+            ]
+
+            transforms_delta = [
+                self._build_transform(*(x[NDOF * i : NDOF * (i + 1)] + delta[j, :]))
+                for i in range(ne)
+                for j in range(NDOF)
+            ]
+
+            for i, ((k1, k2), (ms1, ms2, d)) in enumerate(constraints):
+                t1 = transforms[k1] if k1 not in self.locked else gp_Trsf()
+                t2 = transforms[k2] if k2 not in self.locked else gp_Trsf()
+
+                for m1, m2 in zip(ms1, ms2):
+                    if isinstance(m1, gp_Pnt):
+                        tmp = pt_cost(m1, m2, t1, t2, d)
+
+                        for j in range(NDOF):
+
+                            t1j = transforms_delta[k1 * NDOF + j]
+                            t2j = transforms_delta[k2 * NDOF + j]
+
+                            if k1 not in self.locked:
+                                tmp1 = pt_cost(m1, m2, t1j, t2, d)
+                                rv[k1 * NDOF + j] += (tmp1 - tmp) / DIFF_EPS
+
+                            if k2 not in self.locked:
+                                tmp2 = pt_cost(m1, m2, t1, t2j, d)
+                                rv[k2 * NDOF + j] += (tmp2 - tmp) / DIFF_EPS
+
+                    elif isinstance(m1, gp_Dir):
+                        tmp = dir_cost(m1, m2, t1, t2, d)
+
+                        for j in range(NDOF):
+
+                            t1j = transforms_delta[k1 * NDOF + j]
+                            t2j = transforms_delta[k2 * NDOF + j]
+
+                            if k1 not in self.locked:
+                                tmp1 = dir_cost(m1, m2, t1j, t2, d)
+                                rv[k1 * NDOF + j] += (tmp1 - tmp) / DIFF_EPS
+
+                            if k2 not in self.locked:
+                                tmp2 = dir_cost(m1, m2, t1, t2j, d)
+                                rv[k2 * NDOF + j] += (tmp2 - tmp) / DIFF_EPS
+                    else:
+                        raise NotImplementedError(f"{m1,m2}")
+
+            return rv
+
+        return f, jac
+
+    def solve(self) -> List[Location]:
+
+        x0 = array([el for el in self.entities]).ravel()
+        f, jac = self._cost()
+
+        res = minimize(
+            f,
+            x0,
+            jac=jac,
+            method="BFGS",
+            options=dict(disp=True, ftol=1e-12, maxiter=1000),
```

Comment:
When I run the tests I get the following warning about `ftol`.
```
tests/test_assembly.py::test_constrain
  /home/jwright/Downloads/repos/cadquery/cadquery/occ_impl/solver.py:217: OptimizeWarning: Unknown solver options: ftol
    options=dict(disp=True, ftol=1e-12, maxiter=1000),
```

Reply:
Solved